docs/architecture.md

```diff
@@ -0,0 +1,223 @@
+* Multiple Systems that modify different components can run in parallel without mutexes.
+
+There's a few important operations that need to be addressed however:
+
+* **Deleting entire entities**: Deleting an entity means modifying every component container the entity is associated with. Entities can be marked for deletion with a component, or external queue. 'Deleter' Systems can then delete their respective components near the end of the frame.
+* **Copying entire entities**: It may seem convenient to iterate all of an entity's components and copy them into another entity, but this breaks the rules. Copies of an entity should be created by the same system as the original entity. A dedicated 'mark for copy' and copy systems may be practical depending on the case.
+
+### A solution to all world state?
+
+Entities and Components are powerful, but cannot solve every single storage use case alone. EnTT specifically, only allows one component of a unique type per entity, and only one container of these components (pool) per world (registry). One case may require a variable number of the same component for an entity, or the problem may not be well suited for entities and components at all.
+
+EnTT features customizability and future versions can be more versitile, but either way, ECS is best when combined with other data structures besides just components. For example, queues can be used to pass messages between systems.
+
+## Universe rundown
+
+Universe is designed to support many different kinds of Satellites: Planets, Vehicles, Stars, etc... This can be achieved efficiently through Entities and Components, where Satellites are Entities. When considering keeping groups of Satellites in an N-body simulation, patched conics simulation, or just landed on a planet, it also becomes apparent that some weird scheme is needed to position Satellites.
+
+See [test_application/universes/simple.cpp](https://github.com/TheOpenSpaceProgram/osp-magnum/blob/master/src/test_application/universes/simple.cpp) for how a simple universe can be setup.
+
+### Components
+
+Since Universe uses ECS, Satellites are compositions of Components. Universe Components are simple structs prefixed with **UComp**.
+
+Not much is written yet. **UCompVehicle** contains data for vehicles, **UCompPlanet** contains data for planets, and these can be assigned to Satellites.
+
+### Coordinates
+
+Dealing with space means ridiculously long distances and big numbers. Storing positions in floating point is not ideal, since they lose precision the further the value is from zero. Integers don't suffer from this issue and also run consistently across different processors.
+
+Therefore, signed 64-bit integers was chosen as space coordinates. They are also paired with a power-of-2 scale factor to correlate them to meters, where 2^x unit = 1 meter. A scale factor of 10 means that 1024 units = 1 meter.
```

Comment:
Keep in mind that at least gcc and clang both natively support int128_t.

MSVC has an int128_t that's not a primative, but is a struct with a 64bit upper and lower.

I can't say one way or another what the performance difference is, or if they support this on all platforms or only x86_64

Reply:
Hardware support for 128-bit ints is not that great.

https://stackoverflow.com/questions/34234407/is-there-hardware-support-for-128bit-integers-in-modern-processors 

I remember some discussions about it on discord with Causeless , but personally I though the idea to be a little absurd.

Reply:
We'll probably want hierarchical coordinate systems anyways so I don't see the point of using ultra giant types. I'm actually quite unconvinced that we'll escape having floating point coordinates at the lowest scale, and we'll want to simulate interstellar stuff just for bragging rights and because people will probably want that, so I don't think taking a uniform performance hit for 128bit ints when they're rarely necessary is worth it.